### PR TITLE
feat: task-update-backend

### PR DESCRIPTION
**Pull request by Dawid-Dwk**

### CODE REVIEW

File: packages/backend/src/modules/task/task.controller.ts [hunk 1]
```diff
@@ -32,6 +32,16 @@ class TaskController {
     const taskDto: TaskDTO = this._taskMapper.mapToDto(task)
     return res.status(201).send(taskDto)
   }
+
+  async updateById(req: express.Request, res: express.Response) {
+    const updateTaskCommand: UpdateTaskCommand = req.body
+    const updatedById: Task = await this._taskService.updateById(
+      updateTaskCommand,
+      req.params.taskId
```
Comment (mcwiekala) [inline packages/backend/src/modules/task/task.controller.ts:40]: where is DTO? :p

File: packages/shared/src/api/dto/UpdateTaskCommand.dto.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+export interface UpdateTaskCommand {
+  title: string
+  description: string
+  imageCoverId: string
```
Comment (mcwiekala) [inline packages/shared/src/api/dto/UpdateTaskCommand.dto.ts:4]: what about some optional values?

File: packages/backend/src/modules/dashboard/dashboard.repository.ts [hunk 3]
```diff
@@ -1,27 +1,49 @@
...
     dashboard.columns[columnIndex].tasks.push(savedTask)
     await dashboard.save()
     console.log(
       `Dashboard repository: task ${savedTask} added to idDashboard: ("${createTaskCommand.idDashboard}"), idColumn: ("${createTaskCommand.idColumn}")`
     )
   }
+
+  async updateTaskOnDashboard(updateTaskCommand: UpdateTaskCommand, id: string) {
+    const dashboard = await this._dashboardModel.findById(updateTaskCommand.idDashboard)
+    const idCol = { idColumn: updateTaskCommand.idColumn }
```
Comment (kacper-cyra) [inline packages/backend/src/modules/dashboard/dashboard.repository.ts:31]: Why make it an object, if it's not used anywhere as an object?

File: packages/backend/src/modules/dashboard/dashboard.repository.ts [hunk 4]
```diff
@@ -1,27 +1,49 @@
...
+    const columnIndex = this.findColumnIndex(idCol.idColumn, dashboard.columns)
+    const taskIndex = this.findColumnIndex(id, dashboard.columns[columnIndex].tasks)
+    function updateTask(): void {
```
Comment (kacper-cyra) [inline packages/backend/src/modules/dashboard/dashboard.repository.ts:34]: If it's being used only here, than move it to class private method

File: packages/backend/src/modules/dashboard/dashboard.repository.ts [hunk 5]
```diff
@@ -1,27 +1,49 @@
...
+      const task = dashboard.columns[columnIndex].tasks[taskIndex]
+      if (task.title !== updateTaskCommand.title) {
+        task.title = updateTaskCommand.title
+      }
+      if (task.imageCoverId !== updateTaskCommand.imageCoverId) {
+        task.imageCoverId = updateTaskCommand.imageCoverId
+      }
```
Comment (kacper-cyra) [inline packages/backend/src/modules/dashboard/dashboard.repository.ts:41]: We can simplify it. If we don't plan on preventing update if nothing changed, than we shouldn't care if these values have changed, just overwrite them.

`const task = {...dashboard.columns[columnIndex].tasks[taskIndex], title: updateTaskCommand.title, imageCoverId: updateTaskCommand.imageCoverId}`

File: packages/backend/src/modules/task/task.controller.ts [hunk 6]
```diff
@@ -32,6 +32,16 @@ class TaskController {
     const taskDto: TaskDTO = this._taskMapper.mapToDto(task)
     return res.status(201).send(taskDto)
   }
+
+  async updateById(req: express.Request, res: express.Response) {
+    const updateTaskCommand: UpdateTaskCommand = req.body
+    const updatedById: Task = await this._taskService.updateById(
```
Comment (kacper-cyra) [inline packages/backend/src/modules/task/task.controller.ts:38]: The thing that will be stored here is an updatedTask 😄

File: packages/backend/src/modules/task/task.repository.ts [hunk 7]
```diff
@@ -25,6 +25,21 @@ export class TaskRepository implements Repository<TaskDTO> {
     const task: Task = this._taskModel.findById(id)
     return task
   }
+
+  async updateById(updateTaskCommand: UpdateTaskCommand, id: string): Promise<Task> {
+    const updateById: Task = this._taskModel.updateOne(
```
Comment (kacper-cyra) [inline packages/backend/src/modules/task/task.repository.ts:30]: updatedTask 😛

**Review by kacper-cyra — CHANGES_REQUESTED** [review]
If you need help with anything, ask 😄

File: packages/backend/src/modules/task/task.controller.ts (hunk 1 above)
Comment (Dawid-Dwk) [reply packages/backend/src/modules/task/task.controller.ts:40]: File name updated

File: packages/shared/src/api/dto/UpdateTaskCommand.dto.ts (hunk 2 above)
Comment (Dawid-Dwk) [reply packages/shared/src/api/dto/UpdateTaskCommand.dto.ts:4]: description and imageCoverID change to optional values

Comment (Dawid-Dwk) [reply packages/shared/src/api/dto/UpdateTaskCommand.dto.ts:4]: description and imageCoverID change to optional values

File: packages/backend/src/modules/dashboard/dashboard.repository.ts [hunk 8]
```diff
@@ -1,27 +1,49 @@
-import { CreateTaskCommandDTO } from 'shared'
+import { CreateTaskCommandDTO, UpdateTaskCommand } from 'shared'
 import DashboardModel from './Dashboard'
 import { Task } from '../task/task'
 
 export class DashboardRepository {
   private readonly _dashboardModel = DashboardModel
 
+  private findColumnIndex(id: string, array: []): number {
+    const indexNumber = array.findIndex((x: { _id: string }) => String(x._id) === id)
```
[the remark this answers is not in the document]
Comment (Dawid-Dwk) [reply packages/backend/src/modules/dashboard/dashboard.repository.ts:9]: Done

File: packages/backend/src/modules/dashboard/dashboard.repository.ts (hunk 3 above)
Comment (Dawid-Dwk) [reply packages/backend/src/modules/dashboard/dashboard.repository.ts:31]: Done

File: packages/backend/src/modules/dashboard/dashboard.repository.ts (hunk 4 above)
Comment (Dawid-Dwk) [reply packages/backend/src/modules/dashboard/dashboard.repository.ts:34]: Done

File: packages/backend/src/modules/dashboard/dashboard.repository.ts (hunk 5 above)
Comment (Dawid-Dwk) [reply packages/backend/src/modules/dashboard/dashboard.repository.ts:41]: Done

File: packages/backend/src/modules/task/task.controller.ts (hunk 6 above)
Comment (Dawid-Dwk) [reply packages/backend/src/modules/task/task.controller.ts:38]: Done

File: packages/backend/src/modules/dashboard/dashboard.repository.ts (hunk 5 above)
Comment (Dawid-Dwk) [reply packages/backend/src/modules/dashboard/dashboard.repository.ts:41]: It is nicer solution but for some reason makes the id disappear

File: packages/backend/src/modules/task/task.repository.ts (hunk 7 above)
Comment (Dawid-Dwk) [reply packages/backend/src/modules/task/task.repository.ts:30]: Done

File: packages/backend/src/modules/dashboard/dashboard.repository.ts [hunk 9]
```diff
@@ -1,27 +1,50 @@
-import { CreateTaskCommandDTO } from 'shared'
+import { ObjectId } from 'mongoose'
+import { CreateTaskCommandDTO, UpdateTaskCommandDTO } from 'shared'
 import DashboardModel from './Dashboard'
 import { Task } from '../task/task'
 
 export class DashboardRepository {
   private readonly _dashboardModel = DashboardModel
 
+  // findIndexInDatabase is looking for index by id in an Array of Objects
```
Comment (kacper-cyra) [inline packages/backend/src/modules/dashboard/dashboard.repository.ts:9]: Code should be self explanatory, and comments explaining what the function does are not needed.

**Review by kacper-cyra — DISMISSED** [review]

File: packages/backend/src/modules/dashboard/dashboard.repository.ts [hunk 10]
```diff
@@ -1,27 +1,50 @@
...
+  private findIndexOfDocument(id: string, array: { _id: ObjectId }[]): number {
```
Comment (mcwiekala) [inline packages/backend/src/modules/dashboard/dashboard.repository.ts:10]: index or id? Be consistent

File: packages/backend/src/modules/dashboard/dashboard.repository.ts [hunk 11]
```diff
@@ -1,27 +1,50 @@
...
+    const columnOrTaskIndex = array.findIndex(
+      (columnOrTaskID: { _id: ObjectId }) => String(columnOrTaskID._id) === id
+    )
+    return columnOrTaskIndex
```
Comment (mcwiekala) [inline packages/backend/src/modules/dashboard/dashboard.repository.ts:14]: don.t do code like that. Read about SCR create another metod for task. Be precise

Comment (mcwiekala) [reply packages/backend/src/modules/dashboard/dashboard.repository.ts:14]: https://en.wikipedia.org/wiki/Single-responsibility_principle#:~:text=The%20single%2Dresponsibility%20principle%20(SRP,it%20should%20encapsulate%20that%20part.

File: packages/backend/src/modules/task/task.controller.ts (hunk 1 above)
Comment (mcwiekala) [reply packages/backend/src/modules/task/task.controller.ts:40]: fix it!

File: packages/backend/src/modules/task/task.controller.ts [hunk 12]
```diff
@@ -32,6 +32,16 @@ class TaskController {
...
   }
+
+  async updateById(req: express.Request, res: express.Response) {
+    const updateTaskCommand: UpdateTaskCommandDTO = req.body
+    const updatedTask: Task = await this._taskService.updateById(
+      updateTaskCommand,
+      req.params.taskId
+    )
+    const updatedTaskDto: TaskDTO = this._taskMapper.mapToDto(updatedTask)
+    return res.status(201).send(updatedTaskDto)
```
Comment (mcwiekala) [inline packages/backend/src/modules/task/task.controller.ts:43]: wrong code!
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

File: packages/shared/src/api/dto/UpdateTaskCommand.dto.ts [hunk 13]
```diff
@@ -0,0 +1,9 @@
+export interface UpdateTaskCommandDTO {
+  title: string
```
Comment (mcwiekala) [inline packages/shared/src/api/dto/UpdateTaskCommand.dto.ts:2]: should be also optional

File: packages/backend/src/modules/task/task.service.ts [hunk 14]
```diff
@@ -28,6 +28,12 @@ export class TaskService {
   findById(_id: string): Promise<Task> {
     return this._taskRepository.findById(_id)
   }
+
+  async updateById(updateTaskCommand: UpdateTaskCommandDTO, id: string): Promise<Task> {
```
Comment (mcwiekala) [inline packages/backend/src/modules/task/task.service.ts:32]: which id? Change var name

Comment (mcwiekala) [inline packages/backend/src/modules/task/task.service.ts:32]: which id? Change var name

File: packages/backend/src/modules/task/task.service.ts [hunk 15]
```diff
@@ -28,6 +28,12 @@ export class TaskService {
...
+    const updatedTask: Task = await this._taskRepository.updateById(updateTaskCommand, id)
+    this._boardRepository.updateTaskOnDashboard(updateTaskCommand, id)
```
Comment (mcwiekala) [inline packages/backend/src/modules/task/task.service.ts:34]: rename to: _dashboardRepository

File: packages/backend/src/modules/task/task.service.ts [hunk 16]
```diff
@@ -28,6 +28,12 @@ export class TaskService {
   findById(_id: string): Promise<Task> {
     return this._taskRepository.findById(_id)
   }
+
+  async updateById(updateTaskCommand: UpdateTaskCommandDTO, id: string): Promise<Task> {
+    const updatedTask: Task = await this._taskRepository.updateById(updateTaskCommand, id)
```
Comment (mcwiekala) [inline packages/backend/src/modules/task/task.service.ts:33]: there is no method like: ._taskRepository.updateById

File: packages/backend/src/modules/dashboard/dashboard.repository.ts (hunk 9 above)
Comment (Dawid-Dwk) [reply packages/backend/src/modules/dashboard/dashboard.repository.ts:9]: Deleted

File: packages/backend/src/modules/dashboard/dashboard.repository.ts (hunk 10 above)
Comment (Dawid-Dwk) [reply packages/backend/src/modules/dashboard/dashboard.repository.ts:10]: I looking for index but to do that I need to use id.

File: packages/backend/src/modules/dashboard/dashboard.repository.ts (hunk 11 above)
Comment (Dawid-Dwk) [reply packages/backend/src/modules/dashboard/dashboard.repository.ts:14]: Return name would be confusing but this function has single responsibility. I changed this name.

File: packages/backend/src/modules/task/task.controller.ts (hunk 12 above)
Comment (Dawid-Dwk) [reply packages/backend/src/modules/task/task.controller.ts:43]: Changed 201 -> 200

File: packages/shared/src/api/dto/UpdateTaskCommand.dto.ts (hunk 13 above)
Comment (Dawid-Dwk) [reply packages/shared/src/api/dto/UpdateTaskCommand.dto.ts:2]: Done

File: packages/backend/src/modules/task/task.service.ts (hunk 14 above)
Comment (Dawid-Dwk) [reply packages/backend/src/modules/task/task.service.ts:32]: Updated, id -> taskId

File: packages/backend/src/modules/task/task.service.ts (hunk 15 above)
Comment (Dawid-Dwk) [reply packages/backend/src/modules/task/task.service.ts:34]: Done

File: packages/shared/src/api/dto/UpdateTaskCommand.dto.ts [hunk 17]
```diff
@@ -0,0 +1,9 @@
+export interface UpdateTaskCommandDTO {
```
Comment (mcwiekala) [inline packages/shared/src/api/dto/UpdateTaskCommand.dto.ts:1]: rename to "UpdateTaskCommand" and move to: packages/shared/src/api/command

Comment (Dawid-Dwk) [reply packages/shared/src/api/dto/UpdateTaskCommand.dto.ts]: Done

File: packages/backend/src/modules/task/task.repository.ts [hunk 18]
```diff
@@ -1,4 +1,4 @@
-import { CreateTaskCommandDTO, TaskDTO } from 'shared'
+import { CreateTaskCommandDTO, TaskDTO, UpdateTaskCommand } from 'shared'
```
[the remark this answers is not in the document]
Comment (Dawid-Dwk) [reply packages/backend/src/modules/task/task.repository.ts:1]: we have agreed that this should be ignored

**Review by mcwiekala — CHANGES_REQUESTED** [review]
fix postman

**Review by mcwiekala — APPROVED** [review]

**Review by mcwiekala — APPROVED** [review]

**Pull request merged**